### ISSUE TITLE
Package org.eclipse.xtext.util is accessible from more than one module

### ISSUE THREAD

**Issue by gallandarakhneorg** [eclipse/xtext-core#1338]
When I'm defining my DSL project as a Java module (with Java 11):
```
module io.sarl.lang {
	requires org.eclipse.xtext;
	requires org.eclipse.xtext.util;
        ....
}
```

The Java compiler complains on modules' exports that break the Java standards. The error message is:
```
The package org.eclipse.xtext.util is accessible from more than one module: org.eclipse.xtext, org.eclipse.xtext.util
```

**Comment by cdietrich** [comment]
i dont think this is solvable without breaking apis. otherwise it would be long solved

**Comment by gallandarakhneorg** [comment]
The work around is to remove the `module-info.java` file and put the following line into the MANFIEST.MF file:
```
Automatic-Module-Name: <name-of-the-module>
```
This is not perfect but enables to use Xtext into a real Java modular project until the API of Xtext is cleaned.

**Comment by cdietrich** [comment]
so you would merge the modules ?!? i dont think that would be a good idea

**Comment by gallandarakhneorg** [comment]
This line is in my own project. Nothing changed into the Xtext code.

**Comment by cdietrich** [comment]
i am inclined to close this as wontfix for now. @kthoms @szarnekow what do you think?

**Comment by gallandarakhneorg** [comment]
I think making xtext compliant with JPMS with backward compatibility is still possible but it needs to rethink the plugin architecture.

This work may take time. It is not a problem while the unamed modules are supported by Java (this support may be removed sometimes in the futur according to the Java forums).

I think the plugins that are needed to run the generated Java from an Xtext-based DSL, ie the plugins that are needed into the class and module paths of the Java application that is run by the users of the DSL, eg xbase.lib, should be fully compliant with JPMS. The rest of the libraries that are used to create the DSL compiler and related tools could still be not fully JPMS compliant.

**Comment by gallandarakhneorg** [comment]
I come back to indicates that is is almost impossible to compile a Xtext-based DSL with Java 11 because of the providing of the same package from different Xtext plugins.

For example, after launching MWE2 script for generating my DSL, I obtain the files into `src-gen` of my dsl.ui project. Files that are generated by the MWE2 script cannot be compiled with Java 11. For example, I have the following error on file `LangActivator.java`.
```
import org.eclipse.xtext.util.Modules2;
```
The error message is: `The package io.org.eclipse.xtext.util is accessible from more than one module: org.eclipse.xtext, org.eclipse.xtext.util`.

I have similar problems with `org.eclipse.jface.text` and `org.eclipse.text`.

It means that we cannot move up to Java 11 for our DSL.

**Comment by cdietrich** [comment]
@gallandarakhneorg i assume your "compile" is different from other compiles
compiling with maven, tycho and gradle works fine with what is created by the wizard

**Comment by gallandarakhneorg** [comment]
Dear @cdietrich. Thank you to give me the idea to take a look on the wizard. Since we have created our DSL 7 years ago, we didn't use the wizard before.

Our problem was that we have merge the "DSL.ide" and "DSL.ui" modules into a single module. It seems that the dependency graph of our single module is incompatible with Java 11 module management. We have split our module into a "DSL.ide" and " DSL.ui", and the things are compiling better.

**Comment by cdietrich** [comment]
it should also work if you adapt the workflow to generate ide stuff to ui

**Comment by cdietrich** [comment]
wonder if we could move the Xtext metamodel generates to xtext.util.
and if also moving the genmodel and ecore would be an api breakage
@szarnekow any idea?

**Comment by szarnekow** [comment]
If we move the Xtext package to xtext.util, we'd have a split package `org.eclipse.xtext` between xtext and xtext.util.

That would not be helpful.

It is fair to assume that all clients of `org.eclipse.xtext` do also use `org.eclipse.xtext.util` by transitivity, but not the other way round (though extremely unlikely).
What I *could* imagine: Move all the code from `org.eclipse.xtext.util` to `org.eclipse.xtext` since `xtext.util` is re-exported anyway right now.
Invert the dependency such that `xtext.util` depends on `xtext` (e.g. via import package?) and re-exports the currently available packages under `xtext.util`  but not the bundle org.eclipse.xtext. If that's not possible (technically), xtext.util could also re-export org.eclipse.xtext.

JMS clients would only need to use org.eclipse.xtext but not org.eclipse.xtext.util - org.eclipse.xtext.util would be deprecated.

**Comment by ewillink** [comment]
+1, exactly the same suggestion as I just made on https://bugs.eclipse.org/bugs/show_bug.cgi?id=567161 where the split package causes a signing failure for any user who has checked out / imported org.eclipse.xtext but is not aware of the bad packaging mandating that org.eclipse.xtext.util is in the workspace too.

**Comment by ewillink** [comment]
A similar problem probably exists for the org.eclipse.xtext.generator package that is also in more than one plugin.

**Comment by cdietrich** [comment]
please note the deprecation of org.eclipse.xtext.generator bundle (=> less of a problem there)
